### ISSUE TITLE
[MVP LLaDA2.0] Phase 9.1: Numerical Validation (Incremental Layer-by-Layer)

### ISSUE THREAD

**Issue by AlonKellner-RedHat** [vllm-project/dllm-plugin#42]
## Context

- **Parent:** Issue #39 (Phase 9 orchestration)
- **Follows:** Phase 8 (PR #38) - torch.compile and benchmarking
- **Precedes:** Phase 9.2 (#41) - lm-eval validation
- **Why incremental:** Precise identification of numerical discrepancies at component level

Phase 7+8 (PR #38) delivered functional LLaDA2.0 model with torch.compile, but **zero numerical validation** exists. Current tests validate structure/API contracts, not numerical correctness. We need systematic validation from embeddings → attention → MoE → logits.

---

### In Scope

**8 Incremental Validation Points:**

1. **Embedding Layer** (`embed_tokens`) - Exact match expected
2. **Attention Layer** (`LLaDA2BlockAttention`) - QKV projection, Q/K norm, attention computation, output projection
3. **MoE Layer** (`LLaDA2MoE`) - Router logits, sigmoid, group-limited routing (256→8 groups→top-4→top-k), expert dispatch, routed scaling (2.5x), shared expert, combined output
4. **Decoder Layer** (`LLaDA2DecoderLayer`) - Attention + MoE residuals
5. **Full Transformer Stack** - Sequential layer-by-layer, error accumulation
6. **Final Layer Normalization** (`norm`) - RMSNorm on final hidden states
7. **LM Head Logits** (`lm_head`) - Final logits shape and numerical correctness
8. **E2E Input Tokens → Output Logits** - Full forward pass sanity check

**Critical Sub-Validations:**
- Router precision: FP32 (default) vs BF16 (experimental `VLLM_LLADA2_BF16_ROUTER=1`)
- Group-limited routing logic (8 groups, 32 experts each, top-4 groups, top-k within groups)
- Routed scaling factor (2.5x) validation
- Expert load balancing analysis (no pathological bias)

**Validation Methodology:**
- **Tolerance bounds:**
  - Embeddings: Exact match (deterministic lookup)
  - FP32: atol=1e-5, rtol=1e-4
  - BF16: atol=1e-3, rtol=1e-2
  - Accumulated error: Looser bounds for deep layers (documented)
- **Comparison:** Extract intermediate tensors from HF/SGlang, use `torch.allclose()`
- **Test data:** Fixed seeds, multiple scenarios (single token, short, full block, multi-block)

### Out of Scope

- lm-eval benchmarks (Phase 9.2)
- Categorical/generation task evaluation (Phase 9.2)
- Performance optimization (Phase 8 complete)

---

## Technical Design

**Test location:** `tests/test_llada2_numerical_validation.py`

**Test classes:**
```python
class TestEmbeddingNumericalCorrectness
class TestAttentionNumericalCorrectness
class TestMoENumericalCorrectness
class TestDecoderLayerNumericalCorrectness
class TestTransformerStackNumericalCorrectness
class TestLMHeadNumericalCorrectness
class TestRouterPrecision  # FP32 vs BF16
class TestExpertLoadBalancing
```

**Tolerance framework:** `dllm_plugin/validation_utils.py`
```python
def assert_tensors_close(actual, expected, atol, rtol, name) -> dict
def get_tolerance_bounds(layer_type, dtype, layer_depth) -> tuple[float, float]
```

---

## Dependencies

- **HARD:** Phase 8 complete (PR #38 merged)
- **HARD:** Access to `inclusionAI/LLaDA2.0-mini` weights
- **HARD:** GPU environment (A100 recommended)
- **SOFT:** SGlang installation (optional cross-reference)

---

## Acceptance Criteria

- [ ] All 8 validation points pass within tolerance bounds
- [ ] Router precision comparison (FP32 vs BF16) quantified
- [ ] Group-limited routing logic validated
- [ ] Expert load balancing analysis complete
- [ ] All tolerance bounds documented in `docs/PHASE9.1_NUMERICAL_VALIDATION.md`
- [ ] Test suite merged to `tests/test_llada2_numerical_validation.py`
- [ ] Any discrepancies root-caused and documented

---

## Key Files

- `/dllm_plugin/models/llada2.py` - Model implementation
- `/dllm_plugin/models/llada2_attention.py` - Attention implementation
- `/tests/test_llada2_real_model.py` - Existing test patterns
- `/docs/KNOWN_LIMITATIONS.md` - Router precision tracking
